K7 Login screen

legrego opened the pull request:
This PR implements the K7 Login screen, as described in #20015

@ryankeairns @snide @kobelb 

## Login form
<img width="1310" alt="login" src="https://user-images.githubusercontent.com/3493255/46048123-d9e53d80-c0f5-11e8-9e56-acbe3a8f2b5a.png">

## Invalid credentials
![image](https://user-images.githubusercontent.com/3493255/46147388-621d2d00-c233-11e8-8a75-beba7bb5a2b2.png)


## Session expired
![image](https://user-images.githubusercontent.com/3493255/46147408-6fd2b280-c233-11e8-97c1-8acb7b2e0c06.png)


## No connection to Elasticsearch
![image](https://user-images.githubusercontent.com/3493255/46147515-adcfd680-c233-11e8-8ad4-3a21d2c06b2c.png)


## Insecure connection
![image](https://user-images.githubusercontent.com/3493255/46147433-8416af80-c233-11e8-82dc-6f32a5ec1bd0.png)



Closes #20015

ryankeairns commented:
@legrego this looks really great! 

Here are few items that tidy it up just a bit and make it more accessible. I don't suspect they'll be much effort to change, but regardless, let me know if I can be of assistance or if anything below is unclear.

- [ ] Set the `max-width` of `.loginWelcome__content` to `460px`. This will even out the padding inside the box - currently there is a little extra on the right side due to form inputs having a default max-size of 400px.
- [ ] Place the validation message above/outside of the box and display it inside an EuiCallout (e.g. `<EuiCallout color=“danger” size=“s”>`. It should always be the same width as the form box below. Also, add some spacing between the callout and form box using `<EuiSpacer size="l" />`

For accessibility purposes, we need to add labels and ARIA tags as well. 

- [ ] The snippet below is taken from the *JavaScript* tab in the first section of the [EUI forms page](https://elastic.github.io/eui/#/forms/form-layouts) and should provide what you need.

Note: this will result in the label appearing above the input as opposed to being placeholder text inside the input. You're probably saying "But Ryan, you designed it this way!" - I shouldn't have, the 'always visible' labels are more accessible.

- [ ] For aria tags, we can add `aria-required="true"` and `aria-labelledby="<label id>"` to both inputs.
@snide any others to add? We don't currently have any additional help text here, so describedby seems unnecessary...

```
<EuiFormRow
  label="Text field"
  helpText="I am some friendly help text."
>
  <EuiFieldText name="first" />
</EuiFormRow>
```

Let me know if you have questions!

ryankeairns commented:
@gchaps There are words here (see screenshots above) that I think you'll be interested in reviewing. Thanks for taking a look!

legrego commented:
@ryankeairns thanks for the detailed and easy-to-follow feedback! I addressed most of your comments, but I had a question about your last item:

> Add aria tags. We can add aria-required="true" and aria-labelledby="<label id>" to both inputs.

I added the `aria-required` tags to the inputs, but I was wondering if `aria-labelledby` was required for single input fields contained within a `EuiFormRow`. I was under the impression that the form row would take care of accessibility for the input it owned. I tested adding it to just the username field, but my screen reader didn't call out any difference between the labelledby username field, and the non-labelledby password field. Both fields seemed accessible to me with the screen reader, but I could very well be missing something. Is there a trick to getting the aria tag to work? I tried this:

```javascript
<EuiFormRow label="Username" id="usernameFormRow">
	<EuiFieldText aria-labelledby="usernameFormRow" ...... />
</EuiFormRow>
```

---------

@gchaps for the most part, the messages match what the existing login form used. The one change I made was to the error message that appears when you enter invalid credentials:
Previous: `Oops! Error. Try again.`
New: `Oops! Invalid username/password. Try again.`

I'll happily make any changes you suggest, I just wanted to give you some context 😄

ryankeairns commented:
@legrego thanks for the quick changes!

Good point, I think we can safely do without the `aria-labelledby` since we've already identified the relationship (for screen readers) between the `label` and `input` using `for` and `id`, respectively.

One other accessibility thing we need to tackle is giving focus to the validation message when it appears. Currently, I am unable to reach it by keyboard and it probably ought to be highlighted (or note that it was added) once it appears. I'll do some more tinkering with this in the morning.

snide commented on x-pack/plugins/security/public/views/login/components/basic_login_form.tsx:
```diff
@@ -0,0 +1,184 @@
+            />
+          </EuiFormRow>
+
+          <EuiSpacer />
```
You can remove this one now. `EuiFormRow` comes with its own margins.

snide commented on x-pack/plugins/security/public/views/login/login_react.html:
```diff
@@ -0,0 +1,3 @@
+<div class="container">
```
Don't need this wrapper. It's legacy stuff.

snide commented:
@gchaps can you do an audit of the text that appears in the errors before this gets merged.

ryankeairns commented:
Simply adding `alert="role"` to the `EuiCallout` doesn't seem to do the trick. From what I'm [reading](https://www.w3.org/TR/WCAG20-TECHS/ARIA19.html), the container element needs to be present upon page load for the addition of `alert="role"` to work:

> The error container must be present in the DOM on page load for the error message to be spoken by most screen readers.

It seems we'll need to apply focus using JavaScript instead. 

Also, we should make the message accessible via keyboard. It should be as simple as adding `tabindex = 0` by wrapping the `<EuiKeyboardAcessible>` component around the `<EuiCallOut...>` but that also does not fire, presumably for the same reasons. 

It does work if I try adding `tabindex="0"` as a prop on `<EuiCallOut...>`, however.

snide commented:
Remember, [there's a bug around this in chrome currently](https://bugs.chromium.org/p/chromium/issues/detail?id=882443). Using `role="alert"` or `aria-live` on an item, even if its not in the dom already _usually_ will announce it. You can see this in play currently with Safari (which is the only browser currently not broken with accessibility). We use the same pattern for our toasts.

Remember too, that simply putting an empty `<div aria-live="polite">` around the callout (before it exists would also satisfy stuff in the spec way.

I try to avoid focus on text when we can because most people tend to prefer tab focus to lead to actions. If you don't like the live region stuff You can use the `aria-describedby` [prop](https://github.com/elastic/eui/blob/master/src/components/form/form_row/form_row.js#L126-L128) on `EuiFormRow` to link to the error message if you need. That way it'll get read on focus of the attached input id (but only on focus input, which is kind of lame).

From a user standpoint I think `aria-live` / `role="alert"` is likely the best pattern, because you'll want to know about the change as soon as it happens, without having to focus things. That bug in Chrome is unfortunately just the pits right now and make this stuff tough to test.

ryankeairns commented:
Thanks for explaining that @snide .

@legrego To boil that all down, I think we just need to make two very minor edits:

- [ ] Add `role="alert"` to the `<EuiCallOut...` for login error messages
- [ ] Add `role="status"` to the `<EuiCallOut...` for info messages

Thank you!

timroes commented on x-pack/plugins/security/public/views/login/components/basic_login_form.tsx:
```diff
@@ -0,0 +1,184 @@
+      return (
+        <Fragment>
+          <EuiCallOut
+            size="s"
```
Could we please add `role="alert"` to this element. That will highly increase the accessibility of this error message, since it will automatically be read out to a screenreader user, once it appears.

timroes replied:
Ups, just saw @snide already wrote a comment about that in the PR, sorry :D

timroes commented on x-pack/plugins/security/public/views/login/components/basic_login_form.tsx:
```diff
@@ -0,0 +1,184 @@
+          <EuiFormRow label="Username">
+            <EuiFieldText
+              name="username"
+              title="username"
```
Is there a specific use-case for having a `title` attribute here? If not I would recommend getting rid of this, since it's usually not fulfilling much purpose. Same for the password field.

legrego replied:
no specific user-case for `title`, I'll remove!

timroes commented on x-pack/plugins/security/public/views/login/components/basic_login_form.tsx:
```diff
@@ -0,0 +1,184 @@
+  };
+
+  public componentDidMount() {
+    document.addEventListener('keydown', this.submitOnEnter);
```
Just wrap the form in an actual `<form>` element, make the button below `type="submit"` and make the `this.submit` method be called via `<form onSubmit={this.submit}>` instead. No need for making manual key listeners here.

legrego replied:
Great call, that's much simpler, and the way forms have always worked 👍

timroes commented on x-pack/plugins/security/public/views/login/components/basic_login_form.tsx:
```diff
@@ -0,0 +1,184 @@
+              value={this.state.username}
+              onChange={this.onUsernameChange}
+              disabled={this.state.isLoading}
+              aria-required={true}
```
Please replace `aria-required={true}` by a `required` attribute. [First rule of ARIA usage](https://www.w3.org/TR/using-aria/#rule1) is: never use ARIA if there is native HTML functionality expressing the same. Usually also the native semantic provides a bit more benefits, e.g. in case of `required` the whole form is "marked" as invalid if a required field is missing, so you could actually use a ref of the form element and check it's validity that way. Also a screenreader is that way able to actually determine the forms state, no matter on which form element it is at the moment.

cc @ryankeairns since I think that suggestion was from you

legrego replied:
Changing `aria-required` to `required` marks the html form as invalid, which we are styling with a red background image. Since the user hasn't done anything wrong when first visiting the login page, this feels wrong to me. My first instinct is to override the EUI styles for this, but I wanted to get some other opinions before I took that rather blunt approach


![image](https://user-images.githubusercontent.com/3493255/46077364-95da5300-c15e-11e8-9323-9ceb1891db96.png)

legrego commented:
Github is marking the conversation as outdated, so I'll copy it here:

Changing `aria-required` to `required` marks the html form as invalid, which we are styling with a red background image. Since the user hasn't done anything wrong when first visiting the login page, this feels wrong to me. My first instinct is to override the EUI styles for this, but I wanted to get some other opinions before I took that rather blunt approach


![image](https://user-images.githubusercontent.com/3493255/46077364-95da5300-c15e-11e8-9323-9ceb1891db96.png)

ryankeairns commented:
@legrego Browsers do apply the `invalid` state instantly and the workarounds all have their drawbacks. That said, the `:invalid` styles are probably safe to override in this particular case since they are not ever displayed. As it's designed/built, you can't click on "Log in" until both fields have something in them, so there is never a case where a required field is empty/invalid.

Also, since we don't (rightfully) indicate which specific field was incorrect, there isn't a case where error styles are applied to individual inputs.

timroes commented:
@legrego If we decide that this is design wise not a desired solution, I would be fine attaching `aria-required` for that use case. You should use native attributes when possible, but if that breaks the desired functionality and design we want, I would say it is not possible in that case :-)

ryankeairns commented:
Thanks Larry! I think the last thing we need here is feedback from Gail regarding the copy changes. Other than that, LGTM 👍

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3448/)

gchaps commented:
Here are my thoughts:

**Invalid credentials**

Invalid username or password. Please try again.

**Session expired**

Your session has timed out.  Please log in again.

**No connection to Elasticsearch**

Cannot connect to the Elastiscearch cluster
See the Kibana logs for details and try reloading the page.


**Insecure connection**

A secure connection is required for log in
Contact your system administrator.

legrego commented:
Thanks, @gchaps! I updated the messages and screenshots to match

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3526/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3630/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3640/)

legrego commented:
^ Known flaky test

Jenkins, please test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3653/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3718/)

legrego commented:
Jenkins test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3801/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3808/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3863/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4641/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4646/)

legrego commented:
Unknown CI failure?

Jenkins, test this please

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4666/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5288/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5297/)

legrego commented on x-pack/plugins/security/public/lib/parse_next.ts:
```diff
@@ -6,16 +6,23 @@
   }
 
+  let next: string;
+  if (Array.isArray(query.next) && query.next.length > 0) {
```
This is required by typescript -- the old implementation did not account for multiple `next` query parameters, but TS understands that that's a possibility and requires us to handle that case

legrego commented on x-pack/plugins/security/public/views/login/_login.scss:
```diff
@@ -0,0 +1,74 @@
+}
+
+.loginWelcome::before {
+  // SASSTODO: webpack pipeline isn't setup to handle image urls in SASS yet
```
Not sure -- these were copied over from the "welcome to kibana" feature that Chris worked on, and I assumed (perhaps incorrectly) that the design team was using this tag to track usage. @snide are you tracking an issue for this anywhere?

snide replied:
It's OK to leave this. We're tracking the sasstodos on our own and will handle the removal when the file refs can be used.
